android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/FileBroadcastService.java

```diff
@@ -0,0 +1,236 @@
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import fi.iki.elonen.NanoHTTPD;
+
+/**
+ * Shares URLs via bluetooth.
+ * Also interfaces with PWS to shorten URLs that are too long for Eddystone URLs.
+ * Lastly, it surfaces a persistent notification whenever a URL is currently being broadcast.
+ **/
+@TargetApi(21)
+public class FileBroadcastService extends Service {
+
+    private static final String TAG = "FileBroadcastService";
+    private static final int BROADCASTING_NOTIFICATION_ID = 7;
+    public static final String FILE_KEY = "file";
+    public static final String MIME_TYPE_KEY = "type";
+    public static final int PORT = 1234;
+    private NotificationManagerCompat mNotificationManager;
+    private Handler mHandler = new Handler();
+    private Uri mUri;
+    private String mType;
+    private byte[] mFile;
+    private FileBroadcastServer mFileBroadcastServer;
+
+    /////////////////////////////////
+    // callbacks
+    /////////////////////////////////
+
+    @Override
+    public void onCreate() {
+        super.onCreate();
+    }
+
+    @Override
+    public IBinder onBind(Intent intent) {
+        return null;
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+      if (mFileBroadcastServer != null) {
+        mFileBroadcastServer.stop();
+      }
+      String mType = intent.getStringExtra(MIME_TYPE_KEY);
+      Log.d(TAG, mType);
+      Uri mUri = Uri.parse(intent.getStringExtra(FILE_KEY));
+      Log.d(TAG, mUri.toString());
+      try {
+        mFile = getBytes(getContentResolver().openInputStream(mUri));
+      } catch (FileNotFoundException e) {
+        Log.d(TAG, e.getMessage());
+      } catch (IOException e) {
+        Log.d(TAG, e.getMessage());
+      }
+      mNotificationManager = NotificationManagerCompat.from(this);
+      mFileBroadcastServer = new FileBroadcastServer(PORT);
+      try {
+        mFileBroadcastServer.start();
+        createNotification();
+      } catch (IOException e) {
+        Log.d(TAG, e.getMessage());
+      }
+      sendBroadcast(new Intent("server"));
+      WifiP2pManager mManager = (WifiP2pManager) this.getSystemService(Context.WIFI_P2P_SERVICE);
+      WifiP2pManager.Channel mChannel = mManager.initialize(this, this.getMainLooper(), null);
+      changeWifiName();
+      mManager.discoverPeers(mChannel, new WifiP2pManager.ActionListener() {
+        @Override
+        public void onSuccess() {
+          Log.d(TAG, "discovering");
+        }
+
+        @Override
+        public void onFailure(int reasonCode) {
+          Log.d(TAG, "discovery failed " + reasonCode);
+        }
+      });
+      return START_STICKY;
+    }
+
+    /**
+     * HTTP Server to serve.
+     **/
+    public class FileBroadcastServer extends NanoHTTPD {
+
+        public FileBroadcastServer(int port) {
```

Comment:
It seems strange to pass the port as an argument but grab the file and type from the parent class.  I think it would make sense to pull this out into its own top-level class (and put it in its own file) and have the constructor take all three as arguments.